### PR TITLE
Implement #14

### PR DESCRIPTION
**Pull request by Lol3rrr**
This PR implements #14 

# Open Questions / TODO
I am not set on if this naming and API is what we would want and am open to improvements/suggestions. The general idea seems fine to me and a rather simple change in nature.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -56,10 +56,12 @@
 #![warn(rust_2018_idioms)]
 #![warn(rustdoc::all)]
 
+/*
 #[cfg(any(test, miri))]
 pub(crate) const R: usize = 4;
 #[cfg(not(any(test, miri)))]
 const R: usize = 8;
+*/
```
Comment (jonhoo) [inline src/lib.rs:64]: Just remove this rather than comment it out :)

File: src/lib.rs [hunk 2]
```diff
@@ -111,15 +118,15 @@ impl<T: Clone> Clone for Vc<T> {
...
+impl<T, const R: usize> Default for RawVc<T, R> {
     /// Creates an empty `Vc<T>`.
     #[inline]
     fn default() -> Self {
         Self::new()
     }
 }
 
-impl<T> Vc<T> {
+impl<T, const S: usize> RawVc<T, S> {
```
Comment (jonhoo) [inline src/lib.rs:129]: I'd prefer calling this `R` rather than `S`, just to keep the code consistent.

File: src/lib.rs [hunk 3]
```diff
@@ -87,20 +89,25 @@ pub mod vc {
     pub use super::iter::*;
 }
 
+#[cfg(any(test, miri))]
+pub type Vc<T> = RawVc<T, 4>;
+#[cfg(not(any(test, miri)))]
+pub type Vc<T> = RawVc<T, 8>;
```
Comment (jonhoo) [inline src/lib.rs:95]: Thinking about this some more, let's do:
```rust
const fn default_r() -> usize {
    #[cfg....]
    4
}

pub type Vc<T> = RawVc<T, { default_r() }>;

pub struct RawVc<T, const R: usize = { default_r() }> {
...
}
```
That way people can still make use of the default `R` when using `RawVc` directly.

File: src/lib.rs [hunk 4]
```diff
@@ -87,20 +89,25 @@ pub mod vc {
...
+
 /// A `VecDeque` (and `Vec`) variant that spreads resize load across pushes.
 ///
 /// See the [crate-level documentation] for details.
 ///
 /// [crate-level documentation]: index.html
-pub struct Vc<T> {
+pub struct RawVc<T, const R: usize> {
```
Comment (jonhoo) [inline src/lib.rs:102]: I'm torn on the naming here. I see why you chose `RawVc`, but at the same time "raw" tends to imply unsafe/low-level, which this isn't really. Not sure what's a better name that isn't also way longer though. Maybe `VcC`?

I also think we need to add documentation to the `Vc` type alias saying "this is `RawVc` with the `const` set to a reasonable default value"?

**Review by jonhoo — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (Lol3rrr) [reply src/lib.rs:64]: Yeah that was mostly for testing and keeping the current default settings

File: src/lib.rs (hunk 2 above)
Comment (Lol3rrr) [reply src/lib.rs:129]: That was originally my plan however later in the impl block R is already used for a Range type parameter (see https://github.com/jonhoo/atone/blob/79a65402b8dae6ff484b69042162078811df3bc2/src/lib.rs#L735C9-L735C9 for example)

File: src/lib.rs (hunk 4 above)
Comment (Lol3rrr) [reply src/lib.rs:102]: I feel like `VcC` would be a bit confusing/easy to confuse with normal Vc, so a more distinguished name feels appropriate. Maybe `CustomVc` or something like that?

The Docs in general need a bit more work now again, both on the type alias as well as the [`move_amount`](https://github.com/jonhoo/atone/blob/79a65402b8dae6ff484b69042162078811df3bc2/src/lib.rs#L146C7-L146C7) function, which i am also not sure on the name

File: src/lib.rs (hunk 2 above)
Comment (jonhoo) [reply src/lib.rs:129]: Ah, that's awkward! I think we should rename the `R` for the range type instead then. Make it `Range` — they don't all have to be single-letter names :)

File: src/lib.rs (hunk 4 above)
Comment (jonhoo) [reply src/lib.rs:102]: Let's go with `CustomVc` :+1:

File: src/lib.rs (hunk 3 above)
Comment (Lol3rrr) [reply src/lib.rs:95]: As we seem to support `1.52.1` (judging by the check in the CI).
We cant use this pattern:
```
pub struct RawVc<T, const R: usize = { default_r() }> {
...
}
```
as it fails with `error[E0658]: default values for const generic parameters are experimental`

Comment (jonhoo) [reply src/lib.rs:95]: I'd be okay bumping the MSRV for this :+1:

Comment (Lol3rrr) [reply src/lib.rs:95]: Alright, is the MSRV documented anywhere?

Comment (jonhoo) [reply src/lib.rs]: No, it's just checked in CI and not mentioned anywhere I believe.

File: src/lib.rs [hunk 5]
```diff
@@ -87,20 +82,34 @@ pub mod vc {
...
+    // so thats why the explicit return is needed here
+
+    #[cfg(any(test, miri))]
+    return 4;
+    #[cfg(not(any(test, miri)))]
+    return 8;
+}
+
+// TODO
+// What should the docs say here?
```
Comment (jonhoo) [inline src/lib.rs:96]: I would use the same docs as for `CustomVc`, and then add a line to `CustomVc`'s docs saying "this differs from `Vc` in that you can customize the incremental resize chunk size (`R`)"

File: src/lib.rs [hunk 6]
```diff
@@ -1473,14 +1489,14 @@ impl<T> Vc<T> {
             // but we already have a mechanism for doing so -- old_head!
             let tail_of_old = keep.split_off(at);
             let give_away = mem::replace(&mut self.new_tail, keep);
-            Vc {
+            CustomVc {
```
Comment (jonhoo) [inline src/lib.rs:1492]: I'd just use `Self` here so we don't have to change it ever again :p

File: src/lib.rs [hunk 7]
```diff
@@ -111,15 +120,15 @@ impl<T: Clone> Clone for Vc<T> {
     }
 }
 
-impl<T> Default for Vc<T> {
+impl<T, const R: usize> Default for CustomVc<T, R> {
```
Comment (jonhoo) [inline src/lib.rs:123]: Would you mind checking that `Vc::default()` still works? I worry about the type inference not realizing that the type alias setting `R`, and as a result complaining that the call is ambiguous.

File: src/lib.rs [hunk 8]
```diff
@@ -2046,7 +2062,7 @@ mod tests {
     #[test]
     fn test_split_push() {
         // the code below assumes that R is 4
-        assert_eq!(crate::R, 4);
+        // assert_eq!(crate::R, 4);
```
Comment (jonhoo) [inline src/lib.rs:2065]: I think this should assert on the `move_amount` instead.

File: src/lib.rs (hunk 7 above)
Comment (Lol3rrr) [reply src/lib.rs:129]: I added a new Test case for this [which can be found here](https://github.com/jonhoo/atone/pull/15/commits/357396f7b9f290d16d8a702df105ffd77f0a2ed2#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R2598-R2605) and that seems to work just fine for me, so I think this is not a problem

**Review by jonhoo — APPROVED** [review]
Looks great, thanks!

**Comment by jonhoo** [comment]
Released in 0.3.7 :tada:

**Pull request merged**